docs/sdk/src/reference_docs/defensive_programming.rs

```diff
@@ -0,0 +1,618 @@
+//!   defensive handling of the resulting [`Option`] can occur.
+//! - It may be acceptable to use `except()`, but only if one is completely certain (and has
+//!   performed a check beforehand) that a value won't panic upon unwrapping.  Even this is
+//!   discouraged, however, as future changes to that function could then cause that statement to
+//!   panic.  It is better to ensure all errors are propagated and handled accordingly in some way.
+//! - If a function *can* panic, it usually is prefaced with `unchecked_` to indicate its unsafety.
+//! - If you are writing a function that could panic, [be sure to document it!](https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html#documenting-components)
+//! - Carefully handle mathematical operations.  Many seemingly, simplistic operations, such as
+//!   **arithmetic** in the runtime, could present a number of issues [(see more later in this
+//!   document)](#integer-overflow).
+//!
+//! ### Examples of when to `panic!`
+//!
+//! As you traverse through the codebase (particularly in `substrate/frame`, where the majority of
+//! runtime code lives), you may notice that there occurrences where `panic!` is used explicitly.
+//! This is used when the runtime should stall, rather than keep running, as that is considered
+//! safer. Particularly when it comes to mission critical components, such as block authoring,
+//! consensus, or other protocol-level dependencies, the unauthorized nature of a node may actually
+//! cause harm to the network, and thus stalling would be the better option.
+//!
+//! Take the example of the BABE pallet ([`pallet_babe`]), which doesn't allow for a validator to
+//! participate if it is disabled (see: [frame::traits::DisabledValidators]):
+//!
+//! ```rust
+//! if T::DisabledValidators::is_disabled(authority_index) {
+//! 	panic!(
+//! 		"Validator with index {:?} is disabled and should not be attempting to author blocks.",
+//! 		authority_index,
+//! 	);
+//! }
+//! ```
+//!
+//! There are other such examples throughout various pallets, mostly those who are crucial to the
+//! blockchain's function.
+//!
+//! ### Defensive Traits
+//!
+//! The [`Defensive`](frame::traits::Defensive) trait provides a number of functions, all of which
+//! provide an alternative to 'vanilla' Rust functions, e.g.,:
+//!
+//! - [`defensive_unwrap_or()`](frame::traits::Defensive::defensive_unwrap_or)
+//! - [`defensive_ok_or()`](frame::traits::DefensiveOption::defensive_ok_or)
+//!
+//! The [`Defensive`](frame::traits::Defensive) trait and its companions,
+//! [`DefensiveOption`](frame::traits::DefensiveOption),
+//! [`DefensiveResult`](frame::traits::DefensiveResult) can be used to defensively unwrap
+//! and handle values.  This can be used in place of
+//! an `expect`, and again, only if the developer is sure about the unwrap in the first place.
+//!
+//! Here is a full list of all defensive types:
```

Comment:
We have a `mod defensive` in frame-support (and if not, we can create one), so I would not repeat this list here and instead link to it.